src/libraries/ChainlinkPath.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.17;
+
+import {BytesLib} from "./BytesLib.sol";
+import {ChainlinkOracleImpl} from "../chainlink/oracle/ChainlinkOracleImpl.sol";
+import {AggregatorV3Interface} from "chainlink/interfaces/AggregatorV3Interface.sol";
+
+/// @title Chainlink Path Library
+/// @author 0xSplits
+/// @notice Utilities for converting Chainlink Feed paths to and from bytes
+library ChainlinkPath {
+    /// -----------------------------------------------------------------------
+    /// libraries
+    /// -----------------------------------------------------------------------
+
+    using BytesLib for bytes;
+
+    /// -----------------------------------------------------------------------
+    /// constants
+    /// -----------------------------------------------------------------------
+
+    /// @dev size of one packed encoded feed
+    uint24 private constant PATH_UNIT_SIZE = 25;
+
+    /// -----------------------------------------------------------------------
+    /// functions
+    /// -----------------------------------------------------------------------
+
+    /// @notice get the number of feeds in the path
+    function len(bytes memory path) internal pure returns (uint256) {
+        return path.len(PATH_UNIT_SIZE);
+    }
+
+    /// @notice get feeds from a path (packed encoded bytes)
+    function getFeeds(bytes memory path) internal pure returns (ChainlinkOracleImpl.Feed[] memory feeds) {
+        uint256 length = len(path);
+        feeds = new ChainlinkOracleImpl.Feed[](length);
+        for (uint256 i; i < length;) {
+            feeds[i] = getFeed(path, i);
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @notice get a single feed from a path at a given index
+    function getFeed(bytes memory path, uint256 index) internal pure returns (ChainlinkOracleImpl.Feed memory feed) {
+        uint256 offset = index * PATH_UNIT_SIZE;
+        feed.feed = AggregatorV3Interface(path.toAddress(offset));
+        feed.staleAfter = path.toUint24(offset + 20);
+        feed.decimals = path.toUint8(offset + 23);
+        feed.mul = path.toUint8(offset + 24) == 2;
+    }
+
+    /// @notice get a path (packed encoded bytes) from feeds
+    function getPath(ChainlinkOracleImpl.Feed[] memory feeds) internal pure returns (bytes memory path) {
+        uint256 length = feeds.length;
+        for (uint256 i; i < length;) {
+            path = bytes.concat(
+                path,
+                abi.encodePacked(
+                    feeds[i].feed, feeds[i].staleAfter, feeds[i].decimals, feeds[i].mul ? uint8(2) : uint8(1)
+                )
+            );
+            unchecked {
+                ++i;
+            }
+        }
```

Comment:
i am pretty sure this will allocate a new array on each call; i suppose it doesn't really matter since most of these feeds should be 1-3 in length but ideally we can just allocate the array once and fill it in? i'm pretty sure we can do something like:
```
        uint256 length = feeds.length;
        path = new bytes[](length * PATH_UNIT_SIZE);
        for (uint256 i; i < length;) {
            path[i] = abi.encodePacked(
                    feeds[i].feed, feeds[i].staleAfter, feeds[i].decimals, feeds[i].mul ? uint8(2) : uint8(1)
                );
            unchecked {
                i += PATH_UNIT_SIZE;
            }
        }
```

Reply:
nice

Reply:
also not sure how smart the compiler is here (would have to check the IR output) but can check w tests if you save gas by using a var:
```
        path = new bytes[](length * PATH_UNIT_SIZE);
        Feed feed;
        for (uint256 i; i < length;) {
           feed = feeds[i]
           path[i] = abi.encodePacked(
                    feed.feed, feed.staleAfter, feed.decimals, feed.mul ? uint8(2) : uint8(1)
                );
```

Reply:
realized this won't work while implementing because path[i] refers to a single byte so you have to loop through to add each byte

Reply:
how's your yul? :)